File: examples/.net/cases/tasks/task_move/task_move.ui.case.cs
using AppProtocol.Example.DotNet.Core;
using AppProtocol.Example.DotNet.Core.Shared;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Rendering;

namespace AppProtocol.Example.DotNet.Cases.Tasks.TaskMove;

public sealed class TaskMoveUi : BaseUiCase<TaskMoveUi.ViewState>
{
    private readonly TaskMoveDomain _domainCase = new();
    private bool _moveLocked;

    public TaskMoveUi(UiContext context)
        : base(context, new ViewState())
    {
    }

    public override RenderFragment View() => builder =>
    {
        builder.OpenComponent<TaskMoveViewComponent>(0);
        builder.AddAttribute(1, nameof(TaskMoveViewComponent.Owner), this);
        builder.CloseComponent();
    };

    public override async Task TestAsync()
    {
        var view = View();
        if (view is null)
        {
            throw new InvalidOperationException("test: view must return a visual unit");
        }

        var result = await ServiceAsync(new TaskMoveInput
        {
            TaskId = "task_001",
            TargetStatus = "doing",
        });

        if (result.Task.Status != "doing")
        {
            throw new InvalidOperationException("test: ui service must return the moved task");
        }

        try
        {
            await ServiceAsync(new TaskMoveInput
            {
                TaskId = string.Empty,
                TargetStatus = "done",
            });
        }
        catch
        {
            goto lockChecks;
        }

        throw new InvalidOperationException("test: ui service must reject invalid input");

    lockChecks:
        if (!AcquireMoveLock())
        {
            throw new InvalidOperationException("test: first move lock acquisition must succeed");
        }

        if (AcquireMoveLock())
        {
            throw new InvalidOperationException("test: move lock must reject reentry");
        }

        ReleaseMoveLock();
        if (!AcquireMoveLock())
        {
            throw new InvalidOperationException("test: move lock must be releasable");
        }

        ReleaseMoveLock();
    }

    internal async Task<TaskMoveOutput> ServiceAsync(TaskMoveInput input)
    {
        _domainCase.Validate(input);
        return await RepositoryAsync(input);
    }

    internal async Task<TaskMoveOutput> RepositoryAsync(TaskMoveInput input)
    {
        if (Ctx.Api is null)
        {
            throw new InvalidOperationException("task_move.ui requires ctx.api");
        }

        var response = await Ctx.Api.RequestAsync(new AppHttpRequest(
            "PATCH",
            $"/tasks/{input.TaskId}/status",
            new TaskMoveInput { TargetStatus = input.TargetStatus }));

        var result = Materialize<TaskMoveOutput>(response);
        if (string.IsNullOrWhiteSpace(result.Task.Id))
        {
            throw new InvalidOperationException("task_move.ui received an invalid move response");
        }

        return result;
    }

    internal object ResolveDesignSystem()
    {
        if (Ctx.Packages is null ||
            !Ctx.Packages.TryGetValue("designSystem", out var designSystem) ||
            designSystem is null)
        {
            throw new InvalidOperationException("task_move.ui requires packages.designSystem");
        }

        return designSystem;
    }

    internal TaskCard ResolveTask()
    {
        if (Ctx.Extra is null ||
            !Ctx.Extra.TryGetValue("task", out var task) ||
            task is null)
        {
            throw new InvalidOperationException("task_move.ui requires extra.task");
        }

        return Materialize<TaskCard>(task);
    }

    internal Func<Task>? ResolveOnTaskMoved()
    {
        return Ctx.Extra is not null &&
               Ctx.Extra.TryGetValue("onTaskMoved", out var callback) &&
               callback is Func<Task> onTaskMoved
            ? onTaskMoved
            : null;
    }

    internal bool AcquireMoveLock()
    {
        if (_moveLocked)
        {
            return false;
        }

        _moveLocked = true;
        return true;
    }

    internal void ReleaseMoveLock()
    {
        _moveLocked = false;
    }

    public sealed class ViewState
    {
        public bool Loading { get; init; }
        public string? Error { get; init; }
    }

    public sealed class TaskMoveViewComponent : ComponentBase
    {
        [Parameter]
        public required TaskMoveUi Owner { get; set; }

        private ViewState _state = new();

        private async Task MoveAsync(string nextStatus)
        {
            if (!Owner.AcquireMoveLock())
            {
                return;
            }

            _state = new ViewState { Loading = true };
            await InvokeAsync(StateHasChanged);

            try
            {
                var task = Owner.ResolveTask();
                await Owner.ServiceAsync(new TaskMoveInput
                {
                    TaskId = task.Id,
                    TargetStatus = nextStatus,
                });

                var onTaskMoved = Owner.ResolveOnTaskMoved();
                if (onTaskMoved is not null)
                {
                    await onTaskMoved();
                }

                _state = new ViewState();
            }
            catch (Exception error)
            {
                _state = new ViewState
                {
                    Error = error.Message,
                };
            }
            finally
            {
                Owner.ReleaseMoveLock();
            }

            await InvokeAsync(StateHasChanged);
        }

        protected override void BuildRenderTree(RenderTreeBuilder builder)
        {
            dynamic designSystem = Owner.ResolveDesignSystem();
            var task = Owner.ResolveTask();

            builder.OpenElement(0, "div");
            builder.AddContent(1, (RenderFragment)designSystem.MoveTaskAction(
                this,
                task.Status ?? "todo",
                _state.Loading,
                (Func<string, Task>)MoveAsync));

            if (!string.IsNullOrWhiteSpace(_state.Error))
            {
                builder.OpenElement(2, "div");
                builder.AddAttribute(3, "style", "color: #9a2d1f; font-size: 0.82rem; margin-top: 0.55rem;");
                builder.AddContent(4, _state.Error);
                builder.CloseElement();
            }

            builder.CloseElement();
        }
    }
}

File: examples/java/cases/tasks/task_move/TaskMoveUiCase.java
package cases.tasks.task_move;

import core.BaseUiCase;
import core.shared.AppBaseContext.AppLogger;
import core.shared.AppInfraContracts.AppHttpClient;
import java.util.Map;

public class TaskMoveUiCase extends BaseUiCase<TaskMoveUiCase.ViewState> {
  public interface DesignSystemContract {
    String moveTaskAction(String buttonsHtml);
    String feedback(String type, String message);
  }

  public static class ViewState {
    public boolean loading;
    public String error;
  }

  public TaskMoveUiCase(UiContext ctx) {
    super(ctx, new ViewState());
  }

  @Override
  public String view() {
    DesignSystemContract designSystem = resolveDesignSystem();
    TaskMoveDomainCase.Task task = resolveTask();
    StringBuilder buttons = new StringBuilder();
    for (String status : TaskMoveDomainCase.TASK_STATUS_VALUES) {
      boolean disabled = status.equals(task.status());
      buttons.append("""
          <form method="post" action="/tasks/%s/status">
            <input type="hidden" name="targetStatus" value="%s" />
            <button type="submit" %s>%s</button>
          </form>
          """.formatted(
          task.id(),
          status,
          disabled ? "disabled" : "",
          status
      ));
    }

    String feedback = state.error == null ? "" : designSystem.feedback("error", state.error);
    return "<div>" + designSystem.moveTaskAction(buttons.toString()) + feedback + "</div>";
  }

  @Override
  public void test() throws Exception {
    UiContext testContext = new UiContext();
    testContext.correlationId = "task-move-ui-test";
    testContext.logger = new NoopLogger();
    testContext.api = config -> Map.of(
        "task", Map.of(
            "id", "task_001",
            "title", "Moved task",
            "status", "doing",
            "createdAt", "2026-03-18T12:00:00.000Z",
            "updatedAt", "2026-03-18T12:10:00.000Z"
        )
    );
    testContext.packages.put("designSystem", new DesignSystemContract() {
      @Override
      public String moveTaskAction(String buttonsHtml) {
        return "<div>" + buttonsHtml + "</div>";
      }

      @Override
      public String feedback(String type, String message) {
        return message == null || message.isBlank() ? "" : "<div>" + message + "</div>";
      }
    });
    testContext.extra.put("task", new TaskMoveDomainCase.Task(
        "task_001",
        "Todo task",
        null,
        "todo",
        "2026-03-18T12:00:00.000Z",
        "2026-03-18T12:00:00.000Z"
    ));

    TaskMoveUiCase uiCase = new TaskMoveUiCase(testContext);
    String view = uiCase.view();
    if (view == null || view.isBlank()) {
      throw new IllegalStateException("test: view must return a visual unit");
    }

    TaskMoveDomainCase.TaskMoveOutput result = uiCase.submit(
        new TaskMoveDomainCase.TaskMoveInput("task_001", "doing")
    );
    if (!"doing".equals(result.task().status())) {
      throw new IllegalStateException("test: ui service must return the moved task");
    }

    boolean rejected = false;
    try {
      uiCase.submit(new TaskMoveDomainCase.TaskMoveInput("", "done"));
    } catch (IllegalArgumentException ignored) {
      rejected = true;
    }

    if (!rejected) {
      throw new IllegalStateException("test: ui service must reject invalid input");
    }
  }

  public TaskMoveDomainCase.TaskMoveOutput submit(
      TaskMoveDomainCase.TaskMoveInput input
  ) throws Exception {
    return _service(input);
  }

  protected TaskMoveDomainCase.TaskMoveOutput _service(
      TaskMoveDomainCase.TaskMoveInput input
  ) throws Exception {
    TaskMoveDomainCase domainCase = new TaskMoveDomainCase();
    domainCase.validate(input);
    return _repository(input);
  }

  protected TaskMoveDomainCase.TaskMoveOutput _repository(
      TaskMoveDomainCase.TaskMoveInput input
  ) throws Exception {
    Object response = resolveApiClient().request(Map.of(
        "method", "PATCH",
        "url", "/tasks/" + input.taskId() + "/status",
        "body", Map.of("targetStatus", input.targetStatus())
    ));

    if (!(response instanceof Map<?, ?> mapResponse)) {
      throw new IllegalStateException("task_move.ui received an invalid move response");
    }

    @SuppressWarnings("unchecked")
    Map<String, Object> output = (Map<String, Object>) mapResponse;
    return TaskMoveDomainCase.TaskMoveOutput.fromMap(output);
  }

  private TaskMoveDomainCase.Task resolveTask() {
    Object value = ctx.extra.get("task");
    if (!(value instanceof TaskMoveDomainCase.Task task)) {
      throw new IllegalStateException("task_move.ui requires extra.task");
    }
    return task;
  }

  private DesignSystemContract resolveDesignSystem() {
    Object value = ctx.packages.get("designSystem");
    if (!(value instanceof DesignSystemContract designSystem)) {
      throw new IllegalStateException("task_move.ui requires packages.designSystem");
    }
    return designSystem;
  }

  private AppHttpClient resolveApiClient() {
    if (ctx.api == null) {
      throw new IllegalStateException("task_move.ui requires ctx.api");
    }

    return ctx.api;
  }

  private static final class NoopLogger implements AppLogger {
    @Override
    public void debug(String message, Map<String, Object> meta) {}

    @Override
    public void info(String message, Map<String, Object> meta) {}

    @Override
    public void warn(String message, Map<String, Object> meta) {}

    @Override
    public void error(String message, Map<String, Object> meta) {}
  }
}
